pkg/utils/ldap/identity.go

```diff
@@ -0,0 +1,569 @@
+	var memberValue string
+
+	if strings.ToLower(i.Group.Objectclass) == "posixgroup" {
+		// posixGroup usually means that the member attribute just contains the username
+		memberValue = userEntry.GetEqualFoldAttributeValue(i.User.Schema.Username)
+	} else {
+		// In all other case we assume the member Attribute to contain full LDAP DNs
+		memberValue = userEntry.DN
+	}
+
+	searchRequest := ldap.NewSearchRequest(
+		i.Group.BaseDN, i.Group.scopeVal,
+		ldap.NeverDerefAliases, 0, 0, false,
+		i.getGroupMemberFilter(memberValue),
+		[]string{i.Group.Schema.ID},
+		nil,
+	)
+
+	sr, err := lc.Search(searchRequest)
+	if err != nil {
+		return []string{}, err
+	}
+
+	groups := []string{}
+
+	for _, entry := range sr.Entries {
+		// FIXME this makes the users groups use the cn, not an immutable id
+		// FIXME 1. use the memberof or members attribute of a user to get the groups
+		// FIXME 2. ook up the id for each group
+		groups = append(groups, entry.GetEqualFoldAttributeValue(i.Group.Schema.ID))
+	}
+	return groups, nil
+}
+
+// GetLDAPGroupByID looks up a group by the supplied Id. Returns the corresponding
+// ldap.Entry
+func (i *Identity) GetLDAPGroupByID(log *zerolog.Logger, lc ldap.Client, id string) (*ldap.Entry, error) {
+	var filter string
+	var err error
+	if filter, err = i.getGroupFilter(id); err != nil {
+		return nil, err
+	}
+	return i.GetLDAPGroupByFilter(log, lc, filter)
+}
+
+// GetLDAPGroupByAttribute looks up a single group by attribute (can be "mail", "gid_number",
+// "display_name", "group_name", "group_id"). Returns the corresponding ldap.Entry
+func (i *Identity) GetLDAPGroupByAttribute(log *zerolog.Logger, lc ldap.Client, attribute, value string) (*ldap.Entry, error) {
+	var filter string
+	var err error
+	if filter, err = i.getGroupAttributeFilter(attribute, value); err != nil {
+		return nil, err
+	}
+	return i.GetLDAPGroupByFilter(log, lc, filter)
+}
+
+// GetLDAPGroupByFilter looks up a single group by the supplied LDAP filter
+// returns the corresponding ldap.Entry
+func (i *Identity) GetLDAPGroupByFilter(log *zerolog.Logger, lc ldap.Client, filter string) (*ldap.Entry, error) {
+	searchRequest := ldap.NewSearchRequest(
+		i.Group.BaseDN, i.Group.scopeVal, ldap.NeverDerefAliases, 1, 0, false,
+		filter,
+		[]string{
+			i.Group.Schema.DisplayName,
+			i.Group.Schema.ID,
+			i.Group.Schema.Mail,
+			i.Group.Schema.Groupname,
+			i.Group.Schema.Member,
+		},
+		nil,
+	)
+	log.Debug().Str("backend", "ldap").Msgf("Search %s", i.Group.BaseDN)
+	res, err := lc.Search(searchRequest)
+
+	if err != nil {
+		var errmsg string
+		if lerr, ok := err.(*ldap.Error); ok {
+			if lerr.ResultCode == ldap.LDAPResultSizeLimitExceeded {
+				errmsg = fmt.Sprintf("too many results searching for group '%s'", filter)
+				log.Debug().Str("backend", "ldap").Err(lerr).
+					Str("groupfilter", filter).Msg("too many results searching for group")
+			}
+		}
+		return nil, errtypes.NotFound(errmsg)
```

Comment:
Same as above. Are all errors considered `NotFound` or do we need to log the real error for debugging?